pkg/arvo/lib/lm.hoon

```diff
@@ -0,0 +1,702 @@
+    $(count +(count), w (setr w count wl))
+  ::
+  ::    Inverse of positive definite symmetric matrix, per Bauer & Reinsch 1971.
+  ++  invert
+    ~/  %invert
+    |=  [u=@lms]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    (gauss-elim u)
+  ++  abs
+    |=  [x=@rs]  ^-  @rs
+    ?:  (gth x .0)  x
+    (sub:rs .0 x)
+  ::
+  ::  |x-y| <= tol
+  ++  isclose
+    |=  [x=@rs y=@rs tol=@rs]
+    (lth:rs (abs (sub:rs x y)) tol)
+  ++  gauss-find-next-row
+    |=  [u=@lms i=@ud]  ^-  @ud
+    ~_  leaf+"lagoon-fail"
+    =/  mnu  `(list @ud)`(shape u)
+    =/  mu  `@ud`-:mnu
+    =/  nu  `@ud`+<:mnu
+    =/  ii  i
+    |-  ^-  @ud
+      ?:  (gth ii mu)  i
+      ?.  (isclose (get u ii i) .0 .1e-6)  ii
+    $(ii +(ii))
+  ++  gauss-scale
+    |=  [u=@lms i=@ud]  ^-  @lms
+    ::~&  ["gs" i (unmake:lvs (getr u i)) (get u i i) (unmake:lvs (divs:lvs (getr u i) (get u i i)))]
+    (setr u i (divs:lvs (getr u i) (get u i i)))
+  ++  gauss-replace-down
+    |=  [u=@lms i=@ud]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  mnu  `(list @ud)`(shape u)
+    =/  mu  `@ud`-:mnu
+    =/  nu  `@ud`+<:mnu
+    =/  j  +(i)
+    |-  ^-  @lms
+      ?:  (gth j mu)  `@lms`u
+      ?:  (isclose (get u j i) .0 .1e-6)  $(j +(j), u u)
+      ::need to divide through by first element then subtract row i
+    $(j +(j), u (setr u j (subv:lvs (getr u j) (muls:lvs (getr u i) (get u j i)))))
+  ::
+  ::  Row reduction has two phases:  check for zero in ith column, if so swap.
+  ::  Then replace down and rescale.
+  ++  gauss-row-reduce
+    |=  [u=@lms i=@ud]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  mnu  `(list @ud)`(shape u)
+    =/  mu  `@ud`-:mnu
+    =/  nu  `@ud`+<:mnu
+    =/  i  1
+    |-  ^-  @lms
+      ?:  (gth i mu)  `@lms`u
+      ?.  (isclose (get u i i) .0 .1e-6)  $(i +(i), u (gauss-replace-down (gauss-scale u i) i))
+      =/  ii  (gauss-find-next-row u i)
+    $(i +(i), u (gauss-scale (gauss-replace-down (swapr u i ii) i) i))
+  ++  gauss-replace-up
+    |=  [u=@lms i=@ud]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  mnu  `(list @ud)`(shape u)
+    =/  mu  `@ud`-:mnu
+    =/  nu  `@ud`+<:mnu
+    =/  j  (dec i)
+    |-  ^-  @lms
+      ?:  =(j 0)  `@lms`u
+    $(j (dec j), u (setr u j (subv:lvs (getr u j) (muls:lvs (getr u i) (get u j i)))))
+  ++  gauss-row-replace
+    |=  [u=@lms]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  mnu  `(list @ud)`(shape u)
+    =/  mu  `@ud`-:mnu
+    =/  nu  `@ud`+<:mnu
+    =/  i  1
+    |-  ^-  @lms
+      ?:  (gth i mu)  `@lms`u
+    $(i +(i), u (gauss-replace-up u i))
+  ++  gauss-elim
+    ::~/  %gauss-elim
+    |=  [u=@lms]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  mnu  `(list @ud)`(shape u)
+    =/  mu  `@ud`-:mnu
+    =/  nu  `@ud`+<:mnu
+    ?.  =(mu nu)  !!  :: make sure this is a valid operation (square matrix)
+    =/  i  1
+    =/  u  (augment u)
+    |-  ^-  @lms
+      ?:  (gth i mu)  (gauss-row-replace u)
+    $(i +(i), u (gauss-row-reduce u i))
```

Comment:
```
++  gauss-row-replace
  |=  u=@lms  ^-  @lms
  =+  [m n]=[&1 &2]:(shape u)
  %+  roll  (gulf 1 m)
  |:([i=*@ u=(augment u)] (gauss-row-reduce u i))
::
++  gauss-elim
  |=  u=@lms  ^-  @lms
  =/  n  +:(mate [`&1 `&2]:(shape u))  :: make sure this is a valid operation (square matrix)
  %+  roll  (gulf 1 n)
  |:([i=*@ u=(augment u)] (gauss-row-reduce u i))
```
not sure the rolls in particular are actually less cluttered; I suppose at minimum they're differently cluttered?